hatchet_sdk/clients/workflow_listener.py

```diff
@@ -69,44 +156,69 @@ async def _request(self) -> AsyncIterator[SubscribeToWorkflowRunsRequest]:
             yield request
             self.requests.task_done()
 
+    def cleanup_subscription(self, subscription_id: int):
+        workflow_run_id = self.subscriptionsToWorkflows[subscription_id]
+
+        if workflow_run_id in self.workflowsToSubscriptions:
+            self.workflowsToSubscriptions[workflow_run_id].remove(subscription_id)
+
+        del self.subscriptionsToWorkflows[subscription_id]
+        del self.events[subscription_id]
+
     async def subscribe(self, workflow_run_id: str):
-        self.events[workflow_run_id] = asyncio.Queue()
+        try:
+            # create a new subscription id, place a mutex on the counter
+            await self.subscription_counter_lock.acquire()
+            self.subscription_counter += 1
+            subscription_id = self.subscription_counter
+            self.subscription_counter_lock.release()
 
-        asyncio.create_task(self._init_producer())
+            self.subscriptionsToWorkflows[subscription_id] = workflow_run_id
 
-        await self.requests.put(
-            SubscribeToWorkflowRunsRequest(
-                workflowRunId=workflow_run_id,
+            if workflow_run_id not in self.workflowsToSubscriptions:
+                self.workflowsToSubscriptions[workflow_run_id] = [subscription_id]
+            else:
+                self.workflowsToSubscriptions[workflow_run_id].append(subscription_id)
+
+            self.events[subscription_id] = _Subscription(
+                subscription_id, workflow_run_id
             )
-        )
 
-        while True:
-            event = await self.events[workflow_run_id].get()
-            if event is False:
-                break
-            if event.workflowRunId == workflow_run_id:
-                yield event
-                break  # FIXME this should only break on terminal events... but we're not broadcasting event types
```

Comment:
looks like we lost the fixme. assuming we're only going to send terminal events on this stream?

Reply:
at the moment, we're only sending results over this stream - this should remain true.